### PR TITLE
Add validation error types

### PR DESCRIPTION
**Pull request by ostap0207**
There are obviously a lot of things missing and will be added in future PRs:
1. Mapping of nested errors
2. Using different dry-validation version
3. JSON schema for errors
4. Better overall documentation. Usage guides (for us and integrators)
5. Error URLs

To make this PR mergeable I would like to:
1. ~~Add documentation on specific errors if we agree on the error types~~
   Will do this later as well need to add a documentation library first, but added some more docs to readme.
2. Cover the rest of the code with tests, mostly `to_hash` ✅ 

But it can already be review to make sure that the mapping makes sense.

But everything else will likely be done in separate PRs.

### CODE REVIEW

File: lib/glia/errors/validation_errors.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+          type: BODY_VALIDATION_ERROR,
+          ref: "https://example.com/errors/#{BODY_VALIDATION_ERROR}.html",
+          message: message,
+          error_details: error_details
+        )
+      end
+    end
+
+    # InvalidNumberError
+    class InvalidNumberError < Error
```
Comment (ostap0207) [inline lib/glia/errors/validation_errors.rb:18]: Did not use `Dry::Struct` for these as quite easy to just use normal classes here + didn't want to lock in the project to some specific `dry` version as we also use it for tests.

File: lib/glia/errors/error.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+module Glia
+  module Errors
+    # Base error
+    class Error
+      attr_reader :type, :ref, :message, :error_details
```
Comment (indrekj) [inline lib/glia/errors/error.rb:7]: can it be just `details` or is that already taken?

Comment (ostap0207) [reply lib/glia/errors/error.rb:7]: Yeah, we already use it in most if not all elixir validation errors.

Comment (ostap0207) [reply lib/glia/errors/error.rb:7]: Could be maybe `info`, not sure.

File: lib/glia/errors/error.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+      def error_details_to_hash(details)
+        if details.nil?
+          nil
+        elsif details.is_a?(Hash)
+          details.keys.each_with_object({}) { |field, hash| hash[field] = error_details_to_hash(details[field]) }
+        elsif details.is_a?(Array)
+          details.map { |error| error_details_to_hash(error) }
+        else
+          details.to_hash
```
Comment (SteamRiC) [inline lib/glia/errors/error.rb:30]: should we maybe check if `to_hash` exists before and if it doesn't deafult to `to_s`?

Comment (ostap0207) [reply lib/glia/errors/error.rb:30]: Yeah, I am adding some tests for this right now and will add this soon.

File: lib/glia/errors/mapper.rb [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        'must not include' => InvalidValueError.new,
+        'must be empty' => InvalidValueError.new,
+        'must be filled' => InvalidValueError.new,
+        'cannot be defined' => InvalidValueError.new,
+        # InvalidNumberError or InvalidValueError
+        'must be equal to' => INVALID_NUMBER_OR_VALUE_ERROR,
+        'must not be equal to' => INVALID_NUMBER_OR_VALUE_ERROR,
+        'must be one of' => INVALID_NUMBER_OR_VALUE_ERROR,
+        'must not be one of' => INVALID_NUMBER_OR_VALUE_ERROR,
+        # InvalidLengthError
```
Comment (SteamRiC) [inline lib/glia/errors/mapper.rb:68]: Do we even need this? IMO these could be just `InvalidValueError` as the message `Value does not satisfy required conditions` describes it pretty accurately and the size requirements will be in error details anyway.

Comment (ostap0207) [reply lib/glia/errors/mapper.rb:68]: https://salemove.slack.com/archives/C01H6MXHQ1E/p1607510987005500?thread_ts=1607507821.004500&cid=C01H6MXHQ1E

Comment (ostap0207) [reply lib/glia/errors/mapper.rb:68]: Yes, in theory, everything can be just `invalid_value`, but in future we might want to specify additional information like what was the actual required size, required format, require type. So then we would need to introduce sub_type or something like this, so I think it's better to actually have separate types straight away.

**Comment by ostap0207** [comment]
One idea for future to simplify mapping is to provide `errors.yml` with some special locale, but haven't looked much into that yet.

File: lib/glia/errors/error.rb (hunk 3 above)
Comment (ostap0207) [reply lib/glia/errors/error.rb:37]: https://github.com/salemove/glia-errors-ruby/pull/3/commits/845fc43f19d4869a78cafaeef0f2afafddfbdbf4

**Comment by ostap0207** [comment]
Ready for review

File: lib/glia/errors/error_types.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Glia
+  module Errors
+    BODY_VALIDATION_ERROR = 'body_validation_error'
```
Comment (ostap0207) [inline lib/glia/errors/error_types.rb:5]: Realized that error can also be in the query param. So maybe could be `invalid_input_error` or `input_validation_error`.

I still want to have a separate error for headers (e.g. `invalid_headers_error`), as usually headers is a developer bug.

File: lib/glia/errors/validation_errors.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+      end
+    end
+
+    # InvalidLengthError
+    class InvalidLengthError < Error
+      def initialize
+        super(
+          type: INVALID_LENGTH_ERROR,
+          ref: "https://example.com/errors/#{INVALID_LENGTH_ERROR}.html",
+          message: 'Value length does not match required length'
```
Comment (SteamRiC) [inline lib/glia/errors/validation_errors.rb:45]: maybe `Value does not satisfy required length` like above?
"match required length" could be interpreted as "value must have a certain fixed length" which is not the case for lower than, greater than and so.

File: lib/glia/errors/validation_errors.rb [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        BOOLEAN = 'boolean'
+        ARRAY = 'array'
+        OBJECT = 'object'
+      end
+
+      def initialize(type:)
+        super(
+          type: INVALID_TYPE_ERROR,
+          ref: "https://example.com/errors/#{INVALID_TYPE_ERROR}.html",
+          message: "Value type does not match required type: #{type}"
```
Comment (SteamRiC) [inline lib/glia/errors/validation_errors.rb:88]: maybe drop the "type" in "Value type does not ...". IMO it sounds a bit better

Comment (ostap0207) [reply lib/glia/errors/validation_errors.rb:88]: Hm, how can a value match a type?

File: lib/glia/errors/error.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+          message: message,
+          error_details: error_details_to_hash(@error_details)
+        }.compact
+      end
+
+      private
+
+      def error_details_to_hash(details)
+        if details.nil?
+          nil
```
Comment (SteamRiC) [inline lib/glia/errors/error.rb:29]: maybe also tests for when details are nil and don't match any of the conditions

Comment (ostap0207) [reply lib/glia/errors/error.rb:29]: It's already tested when it's `nil` https://github.com/salemove/glia-errors-ruby/pull/3/files#diff-455cc62ada261acb2a18de652c7149a4b6bb4432ba27b2adae0a7bbeae110dd0R25.

And when it does not satisfy any conditions it here https://github.com/salemove/glia-errors-ruby/pull/3/files#diff-455cc62ada261acb2a18de652c7149a4b6bb4432ba27b2adae0a7bbeae110dd0R17, since it's recursive it will have to process `details` value.

File: lib/glia/errors/validation_errors.rb (hunk 6 above)
Comment (ostap0207) [reply lib/glia/errors/validation_errors.rb:45]: Maybe `Value length does not satisfy required length conditions`

File: README.md [hunk 9]
```diff
@@ -1,13 +1,48 @@
...
-```
+## Require library
+
+```ruby
 require 'glia/errors'
+```
+
+## For Glia developers
+
+### Map from `dry-validation` result
```
Comment (indrekj) [inline README.md:22]: Can you mention this is for 0.x version.

File: README.md [hunk 10]
```diff
@@ -1,13 +1,48 @@
...
+
+```ruby
+schema = Dry::Validation.Schema do
+  # ...
+end
+error = Glia::Errors.from_dry_validation_schema(schema.output, schema.messages)
```
Comment (indrekj) [inline README.md:28]: There should be `result = schema.call(input)` and then you can extract errors from `result`. schema itself does not have any errors.

File: README.md (hunk 9 above)
Comment (ostap0207) [reply README.md:22]: Ah, yeah. I am actually yet to check if there anything drastically different in other versions in terms of error names, but mentioning would be still useful.

File: lib/glia/errors/error.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+      attr_reader :type, :ref, :message, :error_details
+
+      def initialize(type:, ref:, message: nil, error_details: nil)
+        @type = type
+        @ref = ref
+        @message = message
+        @error_details = error_details
+      end
+
+      def to_hash
```
Comment (indrekj) [inline lib/glia/errors/error.rb:16]: in general I think ruby uses `to_h` (e.g. `Array#to_h`).

File: .rubocop.yml [hunk 12]
```diff
@@ -1,3 +1,9 @@
 AllCops:
   TargetRubyVersion: 2.5
   NewCops: enable
+
+Metrics/MethodLength:
+  Max: 80
+
+Metrics/BlockLength:
+  Max: 200
```
Comment (indrekj) [inline .rubocop.yml:9]: with this long block, might as well just disable this lint

File: README.md [hunk 13]
```diff
@@ -1,13 +1,48 @@
...
+response = error.to_hash
+```
+
+## For REST API integrators
+
+```ruby
+glia_error = post_requests('https://api.glia.com/engagement_requests', params)
```
Comment (indrekj) [inline README.md:35]: I know this is a simplified example, but maybe we can use a bit more realistic example. E.g. in most cases you first check status code and then start parsing errors. Also, you likely won't have `glia_error.type` but rather something like `response['type']`.

File: lib/glia/errors.rb [hunk 14]
```diff
@@ -1,6 +1,15 @@
...
+require_relative './errors/error'
+require_relative './errors/error_types'
+require_relative './errors/validation_errors'
+require_relative './errors/mapper'
+
 module Glia
   # Glia REST API errors
-  module Errors; end
+  module Errors
+    def from_dry_validation_schema(output, messages)
```
Comment (indrekj) [inline lib/glia/errors.rb:11]: To allow using this like in the README: `Glia::Errors.from_dry_validation_schema(schema.output, schema.messages)`, this needs to be `def self.from...`

When you call `result = schema.call(input)` then both output and messages are in the same `result` object. So might as well pass only the `result` object here or am I missing something?

File: lib/glia/errors/validation_errors.rb (hunk 7 above)
Comment (SteamRiC) [reply lib/glia/errors/validation_errors.rb:88]: I was thinking along the lines of `type` being a sort of pattern or format that the value must match. but seems a bit strange indeed when I think about it now

File: lib/glia/errors.rb (hunk 14 above)
Comment (indrekj) [reply lib/glia/errors.rb:11]: can you also add test to this (just to make sure it's wired up correctly). I see you have tests for Mapper, so those don't need to be duplicated.

File: lib/glia/errors/mapper.rb [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        'size cannot be less than' => InvalidLengthError.new,
+        'size must be' => InvalidLengthError.new,
+        'size must be within' => InvalidLengthError.new,
+        'length must be' => InvalidLengthError.new,
+        'length must be within' => InvalidLengthError.new,
+        # MissingValueError
+        'is missing' => MissingValueError.new
+      }.freeze
+
+      def self.from_dry_validation_schema(output, messages)
```
Comment (indrekj) [inline lib/glia/errors/mapper.rb:80]: As previously, output and messages are coming from result object: https://github.com/dry-rb/dry-validation/blob/v0.13.3/lib/dry/validation/result.rb#L3-L5 . Maybe `from_dry_validation_result` makes more sense here. Especially because it doesn't always need to be a `Schema`. It could also be created from `Dry::Validation.Params do`. So `_result` seems more generic.

File: lib/glia/errors/validation_errors.rb (hunk 1 above)
Comment (indrekj) [reply lib/glia/errors/validation_errors.rb:18]: :+1: fewer dependencies is better

File: lib/glia/errors/validation_errors.rb (hunk 6 above)
Comment (indrekj) [reply lib/glia/errors/validation_errors.rb:45]: Could we have a more human-friendly message like `User Name has invalid length`? This property is not meant for the computer but for humans' right?

I would assume some clients would just do `body['error_details'].values.flatten.map {|e| e['message']}.join("\n")` for very quick and simple error handling. If everything is `Value` then that's pretty bad imo. Or will implementing this from our side be too complex?

In general, I think it should be very easy for integrators to display errors (e.g. just a list of all the problems). Then for the better UX then can improve and start highlighting each field etc. But something simple to get https://apidock.com/rails/ActiveModel/Errors/full_messages things initially working is useful.

**Review by SteamRiC — APPROVED** [review]
No more comments from me

**Comment by indrekj** [comment]
We also spoke about array element validation. E.g. what happens if third element of an array is invalid? Did we figure out a strategy how to display it? IIRC we wanted to check how dry-validation and others are doing it and think how to do it as well

File: lib/glia/errors/validation_errors.rb (hunk 6 above)
Comment (ostap0207) [reply lib/glia/errors/validation_errors.rb:45]: The errors are not really localized, so the `message` should not be really used with that purpose IMO. The message is for developers to just understand it quickly.

Instead, they should just get error codes and map them to their own error descriptions.

I can see some cases how someone would use the actual error `message` in some not very imporant UI (for example we would probably use it in our legacy admin), so maybe for this reason can add a field name to the message.

**Comment by ostap0207** [comment]
Yes, I added example in the comment some time ago https://salemove.atlassian.net/wiki/spaces/ENG/pages/1565720792/REST+API+Errors?focusedCommentId=1613561875#comment-1613561875. But essentially for array errors we will have just hash with index as a key. This is how `dry-validation` already works. I will be working on mapping of this + test in the next PR.

File: lib/glia/errors/mapper.rb (hunk 15 above)
Comment (ostap0207) [reply lib/glia/errors/mapper.rb:80]: Makes sense, maybe can also just accept `result`.

File: lib/glia/errors/validation_errors.rb (hunk 6 above)
Comment (indrekj) [reply lib/glia/errors/validation_errors.rb:45]: In dry-validation 1.x there's this API:
> contract.call(name: "").errors(full: true).to_h
> # => {:name=>["Name must be filled"]}

Could we use that as the message?

Comment (ostap0207) [reply lib/glia/errors/validation_errors.rb:45]: Not really as we want to have our own messages and don't really on library-specific messages as this will be just very inconsistent. It's not very difficult to construct the messages ourselves.

File: .rubocop.yml (hunk 12 above)
Comment (ostap0207) [reply .rubocop.yml:9]: It was counting `describe` block as well, disabled for test files.

File: lib/glia/errors/validation_errors.rb (hunk 6 above)
Comment (ostap0207) [reply lib/glia/errors/validation_errors.rb]: https://github.com/salemove/glia-errors-ruby/pull/3/commits/23c43706acbd7957f4f8a0ebf68f9873b2aef304

File: lib/glia/errors.rb (hunk 14 above)
Comment (ostap0207) [reply lib/glia/errors.rb]: Changed to accept result.

**Comment by ostap0207** [comment]
Updated

File: lib/glia/errors/validation_errors.rb (hunk 6 above)
Comment (ostap0207) [reply lib/glia/errors/validation_errors.rb]: Better error messages with more details will be in future when we start adding more details to errors, but I think this is good for now.

File: lib/glia/errors.rb [hunk 16]
```diff
@@ -1,6 +1,15 @@
...
+require_relative './errors/error_types'
+require_relative './errors/validation_errors'
+require_relative './errors/mapper'
+
 module Glia
   # Glia REST API errors
-  module Errors; end
+  module Errors
+    def self.from_dry_validation_result(result)
+      Mapper.from_dry_validation_result(result.output, result.messages)
```
Comment (indrekj) [inline lib/glia/errors.rb:12]: pass `result` forward here as well?

Comment (ostap0207) [reply lib/glia/errors.rb:12]: `Mapper.from_dry_validation_result` is used recursively at which point we don't pass result anymore.

File: lib/glia/errors/validation_errors.rb [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+      end
+    end
+
+    # InvalidNumberError
+    class InvalidNumberError < Error
+      def initialize(field:)
+        super(
+          type: INVALID_NUMBER_ERROR,
+          ref: "https://example.com/errors/#{INVALID_NUMBER_ERROR}.html",
+          message: "#{humanize(field)} is invalid"
```
Comment (indrekj) [inline lib/glia/errors/validation_errors.rb:23]: maybe `is not a number`

File: .rubocop.yml (hunk 12 above)
Comment (indrekj) [reply .rubocop.yml:9]: not needed here anymore then or can use lower value?

File: lib/glia/errors.rb (hunk 16 above)
Comment (indrekj) [reply lib/glia/errors.rb:12]: ok

**Review by indrekj — APPROVED** [review]
in general looks fine I think. Before using it though I think we should also add dry-validation 1.x support so we can see if there are any major differences that affect the API design. also this way we don't forget it and don't force us to stay in the lower version.

File: lib/glia/errors/validation_errors.rb (hunk 17 above)
Comment (ostap0207) [reply lib/glia/errors/validation_errors.rb:23]: It is a number, it's just not big or small enough. Later will be more details like, `should be less then`.

**Comment by ostap0207** [comment]
Yeah, makes sense.

File: lib/glia/errors/error.rb [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        elsif details.is_a?(Hash)
+          details.keys.each_with_object({}) do |field, hash|
+            hash[field] = error_details_to_h(details[field])
+          end
+        elsif details.is_a?(Array)
+          details.map { |error| error_details_to_h(error) }
+        elsif details.respond_to?(:to_h)
+          details.to_h
+        else
+          details
```
Comment (urmastalimaa) [inline lib/glia/errors/error.rb:36]: What if details is e.g `StandardError.new("foo")`? What else kind of value could details be here other than string that would be sensible?

File: Rakefile [hunk 19]
```diff
@@ -14,9 +14,9 @@ end
 task default: %i[lint spec]
 
 task :lint do
-  sh "rubocop && rbprettier --check '**/*.rb'"
+  sh "rubocop && rbprettier --check '**/*.rb' --print-width 100"
 end
 
 task :format do
-  sh "rbprettier --write '**/*.rb'"
+  sh "rbprettier --write '**/*.rb' --print-width 100"
```
Comment (urmastalimaa) [inline Rakefile:21]: Shouldn't these go into .prettierrc?

File: lib/glia/errors/mapper.rb [hunk 20]
```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+module Glia
+  module Errors
+    # Maps library specific errors to Glia public errors
+    module Mapper
+      # dry-validation shares the same message for `eql?` validation for string and number,
+      # so we are separeting them ourselves
```
Comment (urmastalimaa) [inline lib/glia/errors/mapper.rb:34]: _separating_

File: test_cases/invalid_length_error_input.json [hunk 21]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (urmastalimaa) [inline test_cases/invalid_length_error_input.json:1]: IMO it would make sense to capture the invalid input and expectations in a single file. The cases are tied through the case name anyway (_invalid_length_string_range_length_1_2_)

Comment (urmastalimaa) [reply test_cases/invalid_length_error_input.json:1]: The test logic is currently a bit too spread around.

File: .rubocop.yml (hunk 12 above)
Comment (urmastalimaa) [reply .rubocop.yml:9]: I think you forgot to remove the rule here though

Comment (ostap0207) [reply .rubocop.yml:9]: Indrek commented about `Metrics/BlockLength` which indeed didn't make sense and now disabled for tests.
`ModuleLength` was added separately, but GitHub showed the old comment incorrectly.

File: test_cases/invalid_length_error_input.json (hunk 21 above)
Comment (ostap0207) [reply test_cases/invalid_length_error_input.json:1]: Yeah can be

File: Rakefile (hunk 19 above)
Comment (ostap0207) [reply Rakefile:21]: Probably better to keep them in `.prettierrc`

File: lib/glia/errors/error.rb (hunk 18 above)
Comment (ostap0207) [reply lib/glia/errors/error.rb:36]: I think any primitive value is OK here (string, number, bool), for others, I think it probably the best to throw the error.

**Comment by ostap0207** [comment]
Updated

**Review by urmastalimaa — APPROVED** [review]
GH didn't notice me of replies or changes for some reason.

**Comment by urmastalimaa** [comment]
Also, Travis has essentially shut down their free CI service... so that's an issue.

**Comment by ostap0207** [comment]
> Also, Travis has essentially shut down their free CI service... so that's an issue.

Thanks, yeah will have a look, it has been quite slow.

**Comment by ostap0207** [comment]
!merge
